packages/strapi-admin/services/permission/engine.js

```diff
@@ -179,48 +181,91 @@ module.exports = conditionProvider => {
       // 1. Replace each condition name by its associated value
       const resolveConditions = map(conditionProvider.get);
 
-      // 2. Only keep the handler of each condition
-      const pickHandlers = map(prop('handler'));
-
-      // 3. Filter conditions, only keep objects and functions
-      const filterValidConditions = filter(isObject);
+      // 2. Filter conditions, only keep objects and functions
+      const filterValidConditions = filter(condition => isObject(condition.handler));
 
-      // 4. Evaluate the conditions if they're a function, returns the object otherwise
+      // 3. Evaluate the conditions if they're a function, returns the object otherwise
       const evaluateConditions = conditions => {
         return Promise.all(
-          conditions.map(cond =>
-            isFunction(cond)
-              ? cond(user, merge(conditionOptions, { permission: cloneDeep(permission) }))
-              : cond
-          )
+          conditions.map(async cond => ({
```

Comment:
you should not return a property called handler if you call it it makes it super hard to understand. 

You could return `{ condition: cond, conditionResult: cond.handler() }`

Reply:
Yeah, probably better to not modify the base handler, you're right